### PR TITLE
Update DFS primitive matching to use ColumnSchema

### LINKED ISSUE #1259: Update DFS's primitive matching to look at Logical Types and Semantic Tags
- With the Woodwork upgrade, we will need to update DFS to look at the primitive's input Logical Types and Semantic Tags, and match them up with the information in the Woodwork DataTable
- This relates to the logic for generating the list of features
- This will use the ColumnSchema present on the primitive inputs/return_types
  - The primitive matching logic should take into account when a primitive return_type is None, and allow DFS stacking to still work
- Completion: Call DFS with features_only=True, and it returns a list of Features (using Woodwork typing to determine the list of features).

### LINKED ISSUE #1263: Replace ignore_variables with ignore_columns
- `ignore_variables` is used to ignore a specific column in an Entity
- With the Woodwork integration, this should be now `ignore_columns`.

### LINKED ISSUE #1264: Replace return_variable_types with return_types to use ColumnSchema
- The woodwork integration means that `return_types` should now be the parameter, which takes in ColumnSchema
- The CFM/DFS logic will need to be updated as well.

### PR DESCRIPTION
**Pull request by thehomebrewnerd**
### Update DFS primitive matching to use ColumnSchema

Closes #1259 
Closes #1263
Closes #1264

This PR updates DFS to use Woodwork ColumnSchemas for primitive matching. In order to generate the same features as before, the logical types for some of the index and foreign key columns in the ecommerce entity set test fixture were updated. These updates break some of our EntitySet tests and can be reverted if we do not need to match the same feature output.

### CODE REVIEW

File: featuretools/synthesis/deep_feature_synthesis.py [hunk 1]
```diff
@@ -789,23 +809,27 @@ def _feature_in_relationship_path(self, relationship_path, feature):
...
                              feature_filter=None):
+
         features = self._features_by_type(all_features=all_features,
-                                          entity=entity,
+                                          dataframe=dataframe,
                                           max_depth=max_depth,
-                                          variable_type=set(input_types))
+                                          # variable_type=set(input_types))
+                                          # TODO: Investigate whether this needs to be a set?
+                                          column_schemas=list(input_types))
```
Comment (thehomebrewnerd) [inline featuretools/synthesis/deep_feature_synthesis.py:831]: ColumnSchema objects don't currently define a hash function, so they can't be placed in a set. Things seem to work fine without using a set here, but if we need or want a set we will need to add `__hash__` to ColumnSchema.

**Review by thehomebrewnerd — COMMENTED** [review]
Just adding a few comments about the changes made in this PR so far.

File: featuretools/synthesis/dfs.py [hunk 2]
```diff
@@ -198,10 +201,11 @@ def dfs(entities=None,
...
 
-        return_variable_types (list[Variable] or str, optional): Types of
-                variables to return. If None, default to
-                Numeric, Discrete, and Boolean. If given as
-                the string 'all', use all available variable types.
+        return_types (list[woodwork.ColumnSchema] or str, optional):
+            List of ColumnSchemas defining the types of
+            columns to return. If None, defaults to returning all
+            numeric, categorical and boolean types. If given as
+            the string 'all', use all available variable types.
```
Comment (rwedge) [inline featuretools/synthesis/dfs.py:208]: instead of "use all available variable types", perhaps "returns all available types"

File: featuretools/tests/synthesis/test_synthesis_utils.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                          primitive_schema=ColumnSchema(logical_type=Integer))
+
+    assert not _schemas_equal(feature_schema=ColumnSchema(logical_type=Integer, semantic_tags={'index'}),
+                              primitive_schema=ColumnSchema(logical_type=Double, semantic_tags={'index'}))
+
+    assert not _schemas_equal(feature_schema=ColumnSchema(logical_type=Integer, semantic_tags={}),
+                              primitive_schema=ColumnSchema(logical_type=Integer, semantic_tags={'index'}))
+
+    assert not _schemas_equal(feature_schema=ColumnSchema(),
+                              primitive_schema=ColumnSchema(logical_type=Integer, semantic_tags={'index'}))
```
Comment (rwedge) [inline featuretools/tests/synthesis/test_synthesis_utils.py:35]: can we also test that just `Integer` and just `'index'` aren't equal to the empty schema

**Comment by rwedge** [comment]
should we make an issue to review DFS behavior once CFM has been implemented?

**Comment by thehomebrewnerd** [comment]
> should we make an issue to review DFS behavior once CFM has been implemented?

@rwedge Are you thinking specifically in terms of how we handle foreign key columns in generating features or something else?

**Comment by rwedge** [comment]
> @rwedge Are you thinking specifically in terms of how we handle foreign key columns in generating features or something else?

Yes and also looking confirming other primitives behave the same and benchmarking how fast DFS runs with the new implementation

**Comment by thehomebrewnerd** [comment]
> > @rwedge Are you thinking specifically in terms of how we handle foreign key columns in generating features or something else?
> 
> Yes and also looking confirming other primitives behave the same and benchmarking how fast DFS runs with the new implementation

Added #1530 to track items related to this.

File: featuretools/tests/synthesis/test_deep_feature_synthesis.py [hunk 4]
```diff
@@ -1520,3 +1536,17 @@ def test_no_transform_stacking():
...
     for feature_name in expected:
         assert feature_with_name(feature_defs, feature_name)
+
+
+def test_builds_seed_features_on_foreign_key_col(es):
+    seed_feature_sessions = ft.Feature(es, 'sessions', 'customer_id', primitive=Negate)
+
+    dfs_obj = DeepFeatureSynthesis(target_dataframe_name='sessions',
+                                   entityset=es,
+                                   agg_primitives=[Mean],
```
[the remark this answers is not in the document]
Comment (thehomebrewnerd) [reply featuretools/tests/synthesis/test_deep_feature_synthesis.py:1546]: Right - likely a copy/paste issue. Removed.

File: featuretools/tests/synthesis/test_synthesis_utils.py (hunk 3 above)
Comment (thehomebrewnerd) [reply featuretools/tests/synthesis/test_synthesis_utils.py:35]: Yep - added!

**Review by rwedge — APPROVED** [review]
LGTM
